### PR TITLE
ath79: add support for PQI Air Pen

### PR DESCRIPTION
**Pull request by srchack**
SoC: AR9330 (or AR9331 revision?)
Ethernet x1, Wireless 2.4G, uSD card slot x1.
USB Power, include Li-Po Battery.

Installation:
Upload the image from telnet command line.
Flash 8MiB. (factory is dual image)
write image is use half 4MiB size from factory.
rewrite 8MiB all size after write openwrt once.

Signed-off-by: YuheiOKAWA <tochiro.srchack@gmail.com>

### CODE REVIEW

File: target/linux/ath79/dts/ar9330_pqi_air-pen.dts [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+				reg = <0x000000 0x040000>;
+				read-only;
+			};
+
+			partition@1 {
+				label = "u-boot-env";
+				reg = <0x040000 0x010000>;
+			};
+
+			art: partition@2 {
```
Comment (mkresin) [inline target/linux/ath79/dts/ar9330_pqi_air-pen.dts]: partition@50000 according the the reg child.

Same issue applies to all following partition nodes.

File: target/linux/ath79/image/tiny.mk [hunk 2]
```diff
@@ -30,3 +30,13 @@ define Device/buffalo_whr-g301n
   SUPPORTED_DEVICES += whr-g301n
 endef
 TARGET_DEVICES += buffalo_whr-g301n
+
+define Device/pqi_air-pen
+  ATH_SOC := ar9330
+  DEVICE_TITLE := PQI Air-Pen
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb2 kmod-usb-storage
+  IMAGE_SIZE := 7680k
+  SUPPORTED_DEVICES += air-pen
```
Comment (mkresin) [inline target/linux/ath79/image/tiny.mk]: need to be `pqi-air-pen` to be compatible with the ar71xx image.

File: target/linux/ath79/image/tiny.mk [hunk 3]
```diff
@@ -30,3 +30,13 @@ define Device/buffalo_whr-g301n
   SUPPORTED_DEVICES += whr-g301n
 endef
 TARGET_DEVICES += buffalo_whr-g301n
+
+define Device/pqi_air-pen
+  ATH_SOC := ar9330
+  DEVICE_TITLE := PQI Air-Pen
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb2 kmod-usb-storage
```
Comment (mkresin) [inline target/linux/ath79/image/tiny.mk]: we usually don't include the `kmod-usb-storage` in the image. It is up to the user to download and install this package. Furthermore, without a filesystem kmod, it doesn't make much sense to include the storage module.

**Review by mkresin — CHANGES_REQUESTED** [review]
Please use ```git commit --amend``` to add changes to your commit. Use ```git push --force``` to overwrite your github repository/branch. The PR will afterwards update automatically. Please don't close this PR and open a new one.

**Comment by mkresin** [comment]
> Upload the image from telnet command line. Flash 8MiB. (factory is dual image) write image is use half 4MiB size from factory. rewrite 8MiB all size after write openwrt once.

If the original layout is a dual image one, how do you ensure that the image is loaded from the beginning of the factory partition instead of the 4 MByte offset (means secondary image is active).

File: target/linux/ath79/base-files/etc/board.d/01_leds [hunk 4]
```diff
@@ -49,6 +49,10 @@ pcs,cr3000)
 	ucidef_set_led_switch "lan3" "LAN3" "pcs:blue:lan3" "switch0" "0x10"
 	ucidef_set_led_switch "lan4" "LAN4" "pcs:blue:lan4" "switch0" "0x02"
 	;;
+pqi,air-pen)
+	ucidef_set_led_wlan "wlan" "WLAN" "$boardname:blue:wlan" "phy0tpt"
+	ucidef_set_led_default "wps" "WPS" "$boardname:blue:wps" "0"
```
[the remark this answers is not in the document]
Comment (mkresin) [reply target/linux/ath79/base-files/etc/board.d/01_leds]: Same applies to the phy0tpt trigger, which is set via dts as well.

**Comment by srchack** [comment]
Thinks.
I done fix code.
change write image from factory.
add boot change procedure.

**Comment by srchack** [comment]
Flash 8MiB. (factory is dual image)
1. write image is use half 4MiB size from factory.
  write initramfs image to mtd6 and mtd7.
2. change next boot setting. next is mtd6 boot.
  echo -ne "\x01" | dd of=/dev/mtdblock8 conv=notrunc bs=1 count=1 seek=0

3. rewrite sysupgrade image.
  change next boot setting. next is mtd4.
  echo -ne "\x00" | dd of=/dev/mtdblock8 conv=notrunc bs=1 count=1 seek=0

OEM mtd.
dev:    size   erasesize  name
mtd0: 00040000 00010000 "u-boot"
mtd1: 00010000 00010000 "u-boot-env"
mtd2: 00010000 00010000 "ART"
mtd3: 00010000 00010000 "NVRAM"
mtd4: 00140000 00010000 "uImage"
mtd5: 00280000 00010000 "rootfs"
mtd6: 00140000 00010000 "uImage2"
mtd7: 00280000 00010000 "rootfs2"
mtd8: 00010000 00010000 "CONF"

**Comment by mkresin** [comment]
I merged a version with the following modification to my staging tree:

- fixed the wrong alphabetical order in `target/linux/ath79/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom` 
- fixed the wrong Device/TARGET_DEVICES in `target/linux/ath79/image/tiny.mk`

FYI: I'm still curious about the arh9k eeprom with a size of 2002 bytes. Never seen an ath9k eeprom of that size and I rather suspect the size is wrong.

Please build an image from staging tree, test it and leave a notw about the results here (still works/fails now).

**Comment by srchack** [comment]
Sorry,

I checked it again.
I retry pullreq from the latest state again.

firmware is not. 
- delete this device code on `target/linux/ath79/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom`

rename is dts file.
- to `ar9330_pqi_air-pen.dts` from `ar9330_pqi-air-pen.dts`

**Comment by mkresin** [comment]
Updated my staging tree. Please check again.

Would have been easy to notice that the ath9k eeprom userspace stuff isn't required at all. The required nuts and bolts are already in the dts ;-).

**Comment by srchack** [comment]
Sorry, I am sorry that it took me a long time.

I confirmed.
There is no problem running firmware on my device.

I reviewed the writing firmware procedure again.

Installation:
uboot access or Upload the image from telnet command line.

You can start from uboot as it is.
If you do not connect to uboot it is troublesome as follows

Upload the image from telnet command line on OEM firmware.
Flash 8MiB. (factory is dual image)
1. write image is use half 4MiB size from factory.
  write initramfs image to mtd6 and mtd7.
  if image size is over mtd6 size, uimage file split and write mtd6 and mtd7.
  ex,
  dd if=openwrt-ath79-tiny-pqi_air-pen-initramfs-kernel.bin of=mtd6.bin bs=1024 count=1280
  dd if=openwrt-ath79-tiny-pqi_air-pen-initramfs-kernel.bin of=mtd7.bin bs=1024 skip=1280

2. change next boot setting. next is mtd6 boot.
  echo -ne "\x01" | dd of=/dev/mtdblock8 conv=notrunc bs=1 count=1 seek=0

3. rewrite sysupgrade image.
  change next boot setting. next is mtd4.
  echo -ne "\x00" | dd of=/dev/mtdblock8 conv=notrunc bs=1 count=1 seek=0
  sysupgrade is reboot, before rewrite next boot setting.

**Comment by adrianschmutzler** [comment]
@srchack @mkresin 

Just stumbled over this device when looking at a differenc ar9330 one. The support in ath79 has some strange settings that need to be fixed and additional room for improvement:

- wmac caldata is set up twice, once with mtd-cal-data in DTS and once with caldata_extract in 10-ath9k-eeprom, both referring to the same location; in ar71xx, only ath79_register_wmac() was used in the mach file. Obviously, one of the statements in ath79 is wrong and needs to be removed.

- For wmac, the DTS sets `mtd-mac-address = <&art 0x2>;` . That's pure nonsense and also does not match anything in ar71xx. It should just be removed.

- The Wifi LED is used for led-boot etc., while the WPS LED is not used at all. I suggest switching boot to WPS led as this would not mix LED use at boot time.

- Why is this assigned to tiny subtarget? Are there any restrictions due to the strange flashing process?

I would be willing to create patches myself based on the feedback.

**Comment by mpratt14** [comment]
@srchack if the flash chip supports 4k sectors you will not be able to save config during sysupgrade

because of this commit
https://github.com/openwrt/openwrt/commit/05d35403b211ccb9bf20e2b3c0b7a30c82c6d6a2

the best way to verify this is to see what the flash is detected as with kernel log
and compare this to linux source code
if the flash supports 4k sectors then this line must be added to device profile in makefile
`BLOCKSIZE := 4k`

**Pull request closed without merging**